In [69]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns

In [70]:
data=pd.read_csv("C:/Users/Navin Kumar/Desktop/coding_jupyter_notebook/DATA/heart-1 (1).csv")

In [71]:
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [72]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [73]:
data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [74]:
data.duplicated().sum()

1

In [75]:
data = data.drop_duplicates()

In [76]:
data.duplicated().sum()

0

In [83]:
y= data['target']
X=data.drop('target',axis=1)

In [84]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [85]:
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2


      3 age: age in years
      4 sex: sex (1 = male; 0 = female)
      9 cp: chest pain type
        -- Value 1: typical angina
        -- Value 2: atypical angina
        -- Value 3: non-anginal pain
        -- Value 4: asymptomatic
     10 trestbps: resting blood pressure (in mm Hg on admission to the hospital)
     12 chol: serum cholestoral in mg/dl
     16 fbs: (fasting blood sugar > 120 mg/dl)  (1 = true; 0 = false)
     17 restecg: resting electrocardiographic result 
     32 thalach: maximum heart rate achieved 33 thalrest: resting heart rate
     38 exang: exercise induced angina (1 = yes; 0 = no)
     40 oldpeak = ST depression induced by exercise relative to rest
     41 slope: the slope of the peak exercise ST segment
        -- Value 1: upsloping
        -- Value 2: flat
        -- Value 3: downsloping
    
     42.Target: Heart Diseases(1=no, 0 = Yes)
    
    

In [12]:
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [88]:
data.shape

(302, 14)

- When we peerform multiple algorithm and we know some algorithm is not going to perform good, thenalgorithm with lowest accuracy  is called `Base Line Algorithm`

`DUMMY CLASSIFIER`

- dummy classifier is used to see how worse any algorithm(how worse the accuracy will come) will perform on the given dataset

In [93]:
#baseline dummy
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.20)


In [94]:
#Standarization and Normalization 
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)


In [97]:
X_train[0]   # 1st row of dataset

array([ 0.92375792,  0.64625619, -0.89617018,  0.57999069, -1.20107791,
       -0.37706776, -1.03174603, -0.27025126,  1.48772757,  2.45430873,
        0.96211004,  1.26399469,  1.08239491])

- `Example UseCASE`
    `from sklearn.dummy import DummyClassifier`
    `import numpy as np`

#### Example training data
    `X_train = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])`
    `y_train = np.array([0, 1, 0, 1])`
    `dummy = DummyClassifier(strategy="most_frequent")`  # Initializes with a strategy
    `dummy.fit(X_train, y_train)`  # Trains the classifier on the training data


### What Happens During fit()?
    -No Learning of Patterns: The fit() method doesn't actually learn complex patterns or relationships between X_train and y_train. Instead, it computes the information needed to apply the selected strategy.

    -For example, with strategy="most_frequent", the classifier will learn that the most frequent class in y_train is 0, and it will always predict 0 for any input during inference.
    
`Predicting with the Trained DummyClassifier:`
    -Once the classifier is fitted, you can use it to make predictions:

- y_pred = dummy.predict(X_train)
  - print(y_pred)

- Output for strategy="most_frequent":
    array([0, 0, 0, 0])

- Why Use a DummyClassifier?
    Baseline Performance: It provides a baseline to compare your more sophisticated models against. If your complex model performs worse than a DummyClassifier, it suggests that the model might not be capturing meaningful patterns.
- Debugging: It helps ensure that your data pipeline and evaluation metrics are working as expected.

In [119]:
#Baseline Dummy
from sklearn.dummy import DummyClassifier
dummy_clf = DummyClassifier()
dummy_clf.fit(X_train,y_train)

DummyClassifier()

In [121]:
X_train

array([[ 0.92375792,  0.64625619, -0.89617018, ...,  0.96211004,
         1.26399469,  1.08239491],
       [-0.03721615,  0.64625619,  0.08110137, ...,  0.96211004,
         0.27495988,  1.08239491],
       [ 2.31183157, -1.54737397,  1.05837293, ..., -0.6260305 ,
        -0.71407492, -0.56859986],
       ...,
       [-0.03721615,  0.64625619,  1.05837293, ..., -0.6260305 ,
        -0.71407492,  1.08239491],
       [ 1.13730771, -1.54737397,  1.05837293, ...,  0.96211004,
        -0.71407492, -0.56859986],
       [ 0.38988344,  0.64625619, -0.89617018, ...,  0.96211004,
         1.26399469,  1.08239491]])

In [122]:
y_train

295    0
248    0
144    1
74     1
190    0
      ..
258    0
128    1
70     1
39     1
236    0
Name: target, Length: 241, dtype: int64

In [126]:
y_pred=dummy_clf.predict(X_test) #You typically want to predict on X_test because this data was not seen by the model during training. This provides a realistic evaluation of the model's performance.
y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

In [113]:
dummy_clf.score(X_test,y_test)

0.5737704918032787

### Logistic Regression

In [128]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()
logreg.fit(X_train,y_train)


LogisticRegression()

In [143]:
y_pred_log= logreg.predict(X_test)
y_pred_log

array([1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1], dtype=int64)

In [144]:
logreg.score(X_test,y_test)  #testing accuracy

0.8524590163934426

In [146]:
logreg.score(X_train,y_train) #testing accuracy 

0.8298755186721992

When testing accuracy is higher than training accuracy, it’s an unusual situation but can occur under certain conditions. Here are a few potential reasons for this:

- `Data Leakage:` There might be data leakage where the testing data inadvertently includes information from the training data. This would lead to an inflated testing accuracy.

- `Small Sample Size:` If the training set is very small, the model might not have learned enough to perform well on it, but it could perform surprisingly well on the test set if the test set happens to be more aligned with the model's learned patterns.

- `Random Fluctuations:` Sometimes, due to randomness in data splitting or the inherent variability in small datasets, you might see fluctuations where the test accuracy is temporarily higher than the training accuracy.

- `Overfitting Not Present:' In some cases, the model might not be overfitting, and the test data might be less noisy or more representative than the training data.

- `Evaluation Metrics Issue:` Make sure you’re correctly splitting the data and using the right metrics. Sometimes mistakes in preprocessing or evaluation can lead to misleading results.

#### Model performance

In [150]:
from sklearn import metrics
from sklearn.metrics import confusion_matrix
lr_acc_score = metrics.accuracy_score(y_test,y_pred_log)
lr_acc_score

0.8524590163934426

In [151]:
lr_f1_score = metrics.f1_score(y_test,y_pred_log)
lr_f1_score

0.8732394366197183

In [152]:
lr_conf_matrix = confusion_matrix(y_test,y_pred_log)
lr_conf_matrix

array([[21,  5],
       [ 4, 31]], dtype=int64)

In [158]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred_log))

              precision    recall  f1-score   support

           0       0.84      0.81      0.82        26
           1       0.86      0.89      0.87        35

    accuracy                           0.85        61
   macro avg       0.85      0.85      0.85        61
weighted avg       0.85      0.85      0.85        61



### Decision Tree Classifier

In [160]:
from sklearn.tree import DecisionTreeClassifier 
tree= DecisionTreeClassifier()
tree.fit(X_train,y_train)

DecisionTreeClassifier()

In [161]:
y_pred=tree.predict(X_test)
tree_model_accuracy = metrics.accuracy_score(y_test,y_pred)
tree_model_accuracy

0.7868852459016393

In [162]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.71      0.85      0.77        26
           1       0.87      0.74      0.80        35

    accuracy                           0.79        61
   macro avg       0.79      0.79      0.79        61
weighted avg       0.80      0.79      0.79        61



In [163]:
from sklearn.ensemble import RandomForestClassifier
rfc=RandomForestClassifier()
rfc.fit(X_train,y_train)

RandomForestClassifier()

In [164]:
y_pred_rfc = rfc.predict(X_test)
y_pred_rfc

array([1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1], dtype=int64)

In [165]:
accuracy = metrics.accuracy_score(y_test,y_pred_rfc)
accuracy

0.8688524590163934

In [167]:
print(classification_report(y_test,y_pred_rfc))

              precision    recall  f1-score   support

           0       0.88      0.81      0.84        26
           1       0.86      0.91      0.89        35

    accuracy                           0.87        61
   macro avg       0.87      0.86      0.86        61
weighted avg       0.87      0.87      0.87        61

